File: desktop/Netsira.Desktop/Core/RfCalculators.cs
// SPDX-License-Identifier: AGPL-3.0-or-later
namespace Netsira.Desktop.Core;

public static class RfCalculators
{
    public static double FsPlDb(double distanceKm, double frequencyMHz)
    {
        if (distanceKm <= 0) throw new ArgumentOutOfRangeException(nameof(distanceKm));
        if (frequencyMHz <= 0) throw new ArgumentOutOfRangeException(nameof(frequencyMHz));
        return 32.44 + 20 * Math.Log10(distanceKm) + 20 * Math.Log10(frequencyMHz);
    }

    public static double FresnelRadiusMeters(double d1Km, double d2Km, double frequencyGHz)
    {
        if (d1Km <= 0 || d2Km <= 0) throw new ArgumentOutOfRangeException(nameof(d1Km));
        if (frequencyGHz <= 0) throw new ArgumentOutOfRangeException(nameof(frequencyGHz));
        var wavelength = 0.299792458 / frequencyGHz;
        var d1 = d1Km * 1000.0;
        var d2 = d2Km * 1000.0;
        return Math.Sqrt(wavelength * d1 * d2 / (d1 + d2));
    }

    public static double EirpDbm(double txPowerDbm, double antennaGainDbi, double cableLossDb = 0) =>
        txPowerDbm + antennaGainDbi - cableLossDb;

    public static double DbmToMilliwatts(double dbm) => Math.Pow(10, dbm / 10.0);

    public static double MilliwattsToDbm(double milliwatts)
    {
        if (milliwatts <= 0) throw new ArgumentOutOfRangeException(nameof(milliwatts));
        return 10 * Math.Log10(milliwatts);
    }
}

File: android/app/src/main/java/io/github/shadowur0/netsira/core/RfCalculators.kt
// SPDX-License-Identifier: AGPL-3.0-or-later
package io.github.shadowur0.netsira.core

import kotlin.math.log10
import kotlin.math.pow
import kotlin.math.sqrt

object RfCalculators {
    fun fsplDb(distanceKm: Double, frequencyMHz: Double): Double {
        require(distanceKm > 0 && frequencyMHz > 0)
        return 32.44 + 20 * log10(distanceKm) + 20 * log10(frequencyMHz)
    }
    fun fresnelRadiusMeters(d1Km: Double, d2Km: Double, frequencyGHz: Double): Double {
        require(d1Km > 0 && d2Km > 0 && frequencyGHz > 0)
        val wavelength = 0.299792458 / frequencyGHz
        val d1 = d1Km * 1000
        val d2 = d2Km * 1000
        return sqrt(wavelength * d1 * d2 / (d1 + d2))
    }
    fun eirpDbm(txPowerDbm: Double, antennaGainDbi: Double, cableLossDb: Double = 0.0) =
        txPowerDbm + antennaGainDbi - cableLossDb
    fun dbmToMilliwatts(dbm: Double) = 10.0.pow(dbm / 10.0)
}
